clang/lib/Lex/Lexer.cpp

```diff
@@ -2068,7 +2068,8 @@ bool Lexer::LexNumericConstant(Token &Result, const char *CurPtr) {
   }
 
   // If we have a digit separator, continue.
-  if (C == '\'' && (LangOpts.CPlusPlus14 || LangOpts.C23)) {
+  if (C == '\'' &&
+      (LangOpts.CPlusPlus14 || LangOpts.C23 || ParsingPreprocessorDirective)) {
```

Comment:
Why is `ParsingPreprocessorDirective` used here?

I think I would prefer introducing new `LangOptions::AllowLiteralDigitSeparator` member that would be set to true whenever `CPlusPlus14 || C23` evaluates to `true` and that could be set to `true` explicitly from the scanner. WDYT?

Reply:
I'm not certain that's a pattern we should get into -- we don't want a `LangOption` per language feature shared between languages unless it's something the user can enable or disable themselves via the command line.

Reply:
I'd argue that `LangOptions` are options internal to the compiler and not necessarily exposed via `-cc1` (let alone the driver). The tooling use case is relevant too, IMO.

Reply:
There's not a one-to-one mapping of `LangOptions` to frontend of driver options, yes. But I think it's a stretch to argue that the standard mode tracked by `LangOptions` is not exposed via flags.

I don't think per-feature `LangOptions` is viable. There's digit separators, hex float support, binary literal support, different literal suffixes, various keywords, etc and that doesn't seem likely to scale well. Also, it makes it seem like we support a la carte language features when we don't -- we intentionally don't want to let users disable standard features (in general; exceptions exist) and so that approach gives the impression contributors need to check both language mode and feature availability which is a maintenance concern.

I'm still wondering about the question I asked here: https://github.com/llvm/llvm-project/pull/93753#issuecomment-2173666602

Reply:
> But I think it's a stretch to argue that the standard mode tracked by `LangOptions` is not exposed via flags.

Not sure I understand. I'm suggesting that in the future, parts of the lexer behavior could be controlled in a way that's hidden from users. And I'm saying that we are already hiding existing `LangOptions` fields from users (e.g. `LangOptions::CompilingPCH`).

> I don't think per-feature `LangOptions` is viable. There's digit separators, hex float support, binary literal support, different literal suffixes, various keywords, etc and that doesn't seem likely to scale well.

This doesn't need to scale. We'd only do this for options that control lexing. FWIW, code that says "literal digit separators are enabled in C++14+, C23+, or when explicitly allowed" seems more logical to me than checking `ParsingPreprocessorDirective`, which doesn't really explain why we're suddenly allowing digit separators.

> Also, it makes it seem like we support a la carte language features when we don't -- we intentionally don't want to let users disable standard features (in general; exceptions exist) and so that approach gives the impression contributors need to check both language mode and feature availability which is a maintenance concern.

I think documenting the cases where we introduce this extra knob should pretty much clear up any confusion and prevent contributors cargo culting that approach to options that don't need it.

> I'm still wondering about the question I asked here: [#93753 (comment)](https://github.com/llvm/llvm-project/pull/93753#issuecomment-2173666602)

Are there any comments that weren't addressed by [my reply](https://github.com/llvm/llvm-project/pull/93753#issuecomment-2204578263)?

Reply:
>> I'm still wondering about the question I asked here: https://github.com/llvm/llvm-project/pull/93753#issuecomment-2173666602
> Are there any comments that weren't addressed by https://github.com/llvm/llvm-project/pull/93753#issuecomment-2204578263?

Oh, thank you for pointing that out, somehow I missed that you commented (sorry)! I'll keep the discussion over there.

> This doesn't need to scale. We'd only do this for options that control lexing. FWIW, code that says "literal digit separators are enabled in C++14+, C23+, or when explicitly allowed" seems more logical to me than checking ParsingPreprocessorDirective, which doesn't really explain why we're suddenly allowing digit separators.

Yeah, this was the bit I think I was missing from the earlier discussion. Making sure I understand correctly, it's not *all* tokens we need to care about, it's only ones that matter for scanning of dependencies, which basically means punctuators, preprocessor directive keywords, and literals? Or do you mean "control lexing" in terms of what is or isn't a keyword (e.g., `_Atomic` is a keyword in C and C++ except when OpenCL is enabled)?

Reply:
Can we introduce a `ScanningDependencies` flag somewhere (in `LangOpts`) ? that seems to covber the feedback in the previous PR and would let us handle future such cases

Reply:
I would find that more clear than the current approach, WDYT @jansvoboda11?

Reply:
We've been trying to avoid specializing Clang itself for dependency scanning. Instead, we have introduced generic feature flags that the scanner uses to tweak the compiler behavior. I think this is better for layering and also forces us to name those features, making the code self-documenting. People working on Clang don't have to wonder what's dependency scanning and why it needs to lex numeric literals differently; they just see there's a feature flag named `AllowLiteralDigitSeparator` that just makes sense in that context. You can still dive deeper and find that it's the scanner that enables that feature of course.

So I'm not a fan of introducing `LanguageOpts::ScanningDependencies`.

Reply:
> We've been trying to avoid specializing Clang itself for dependency scanning

Except the only reason for these flags to exists is because of dependency scanning.
The sad reality is that there are preprocessor divergence across language modes and the only way we can avoid that is by creating a dialect.

I do agree with you that we should 1/ Try to minimize such divergences and 2/ Document the preprocessor as understood by the dependency scanner.

As this PR shows, Clang cannot ignore the existence of this dialect, and I think being coy about it isn't in fact making maintenance either: ultimately, there is a tension between C and C++ preprocessor evolving and sometimes diverging, and tooling needing to pretend the preprocessor is not affected by language mode.

Clang needs to deal with that and I don't think it's wide to bury that in the driver or to add lots of very targeted flags that will only ever make sense in context of dependency scanning.

Reply:
> We've been trying to avoid specializing Clang itself for dependency scanning. Instead, we have introduced generic feature flags that the scanner uses to tweak the compiler behavior.

That's still specializing Clang for dependency scanning though. We don't typically *want* generic feature flags that suggest an a la carte approach to the language features, so those are only introduced for dependency scanning unless we happen to have a feature flag allowing the feature to be enabled or disabled by the user.

>  People working on Clang don't have to wonder what's dependency scanning and why it needs to lex numeric literals differently; they just see there's a feature flag named `AllowLiteralDigitSeparator` that just makes sense in that context. You can still dive deeper and find that it's the scanner that enables that feature of course.

That's actually the problem though -- folks are telling you that we feel the need to dive deeper when seeing those sort of flags because we need to figure out under what conditions the flag is or is not set in order to determine whether we have appropriate test coverage for changes. Using your example, `AllowLiteralDigitSeparators` instead of looking at the language standard means that reviewers have to ask questions like "so what happens when SYCL requires C++20 but `AllowLiteralDigitSeparators` is set to false?" or similar. Tying language features to language modes makes this far easier for Clang developers to reason about -- C++20 mode implies C++14 mode and so digit separators always work.

Reply:
@AaronBallman @jansvoboda11 @cor3ntin wanted to continue more on this after a long gap I took. How can we reach a consensus? Unfortunately, I am pretty new to Clang and do not understand the implications of making a choice one way or another (which is what this thread is about, so thank you all for nice discussion!).

Reply:
@jansvoboda11, @cor3ntin -- has your opinion changed at all? I'm still pretty strongly opposed to adding feature-specific flags to `LangOptions` for supporting this in dependency scanning. CC @erichkeane @Sirraide @Fznamznon for more opinions as well.

Reply:
I feel like another issue is that just accepting separators uncoditionally in preprocessor directives (i.e. even outside of dependency scanning) would technically be a language extension... doesn’t that mean that we’d technically have to add a compatibility warning? Because we’re usually pretty strict about doing that for things that are effectively a compiler extension.

Also, I’m not really that familiar w/ the dependency scanner, but I’m still confused as to why passing down the `LangOpts` isn’t an option: in #93753, it was pointed out that that would impair caching the results of the scan—but isn’t it already the case that e.g. what files something depends on is dependent on the language standard? E.g. what if I write:
```c++
#if __cplusplus >= WhateverTheValueForC++17IsAgain
#    include "foo.hh"
#else
#    include "bar.hh"
#endif
```

Does it just note that the file might depend on both `foo.hh` and `bar.hh`? Because if so, can’t we just set the language standard in the lang opts for the dependency scanner to the latest e.g. C++ standard? Because that would also fix this issue.

If we really can’t pass down the original lang opts for whatever reason, then I’m in favour of adding a language option either for `DependencyScanning` or for just literal separators.

> I don't think per-feature LangOptions is viable. There's digit separators, hex float support, binary literal support, different literal suffixes, various keywords, etc and that doesn't seem likely to scale well.

I mean, we don’t need a language option for *every* feature either, but adding them on an as-needed basis doesn’t seem completely untenable. We already have, like, over 300 of them, so I feel like we’re past the point where adding more of them would really make that big of a difference.

That said, I can’t seem to decide which of the following options would be the best (other than passing down the original lang opts):
1. Adding a language option for literal separators (or for an individual feature in general) solves the problem of a ‘dependency scanner’ flag potentially being confusing—and, more generally speaking, also has the nice side-effect that you would no longer have to remember the full list of language dialects which support any one feature—except that as Aaron pointed out, I don’t think a dependency scanner flag would be that confusing (‘language dialect X (which dependency scanning would be at that point) needs to do a weird thing’ isn’t really that unusual imo), and also, you still need to know what dialect(s) introduce certain features so you know when to emit compatibility diagnostics.
2. Adding a language option for dependency scanning means we only need a single one to account for this and however many other problems we might run into wrt language dialects causing a problem w/ dependency scanning, but it also means that now every place that checks whether a feature is supported (possibly only in the lexer/preprocessor; again, I don’t know the dependency scanner that well...) also needs to check if we’re performing dependency scanning.

Either way, I’d definitely prefer either option over just... allowing this in preprocessor directives irrespective of the language dialect.

Reply:
> I feel like another issue is that just accepting separators uncoditionally in preprocessor directives (i.e. even outside of dependency scanning) would technically be a language extension... doesn’t that mean that we’d technically have to add a compatibility warning? Because we’re usually pretty strict about doing that for things that are effectively a compiler extension.

Sure, but we don't have to expose that language extension to users, it can remain internal to the compiler.

> Also, I’m not really that familiar w/ the dependency scanner, but I’m still confused as to why passing down the `LangOpts` isn’t an option: in #93753, it was pointed out that that would impair caching the results of the scan—but isn’t it already the case that e.g. what files something depends on is dependent on the language standard? E.g. what if I write:
> 
> ```c++
> #if __cplusplus >= WhateverTheValueForC++17IsAgain
> #    include "foo.hh"
> #else
> #    include "bar.hh"
> #endif
> ```
> 
> Does it just note that the file might depend on both `foo.hh` and `bar.hh`? Because if so, can’t we just set the language standard in the lang opts for the dependency scanner to the latest e.g. C++ standard? Because that would also fix this issue.

The scanner works in two steps:

1. The virtual file system "minimizes" source files into sequence of tokens that are relevant for dependency discovery. Your code fragment would be transformed into something like `[if, ident, gte, ident, then, include, string, else, include, string, endif]`. 
2. The scanner instance takes those tokens and the TU `LanguageOptions` and runs full preprocessor on those tokens. This means that the condition should evaluate correctly.

Having (1) be context-free is important for performance.

On high level, the goal of #93753 is to respect `LanguageOptions` in the scanner. My argument is that we should be able to do so by representing tokens in the stream produced by (1) such that we're still able to emit context-aware, `LanguageOptions`-driven diagnostics in (2).

That solves the issue outlined by #93753 without impacting the performance.

> I mean, we don’t need a language option for _every_ feature either, but adding them on an as-needed basis doesn’t seem completely untenable.

+1

Reply:
> Sure, but we don't have to expose that language extension to users, it can remain internal to the compiler.

Yeah, that’s fine. What I meant was if we were to just use `ParsingPreprocessorDirective`, we would be exposing it to users, so that’s probably not the way to go imo.

> The scanner works in two steps:
> 
> 1. The virtual file system "minimizes" source files into sequence of tokens that are relevant for dependency discovery. Your code fragment would be transformed into something like `[if, ident, gte, ident, then, include, string, else, include, string, endif]`.
> 2. The scanner instance takes those tokens and the TU `LanguageOptions` and runs full preprocessor on those tokens. This means that the condition should evaluate correctly.

I see; in that case being more permissive in what we accept during step 1 makes perfect sense imo. I don’t think there are *that* many language features that affect this step either since that’s basically just preprocessor directives from what I gather, so adding language options for whatever features we need to enable for this seems reasonable.

Reply:
> Sure, but we don't have to expose that language extension to users, it can remain internal to the compiler.

This still doesn't address my concerns from earlier about how it impacts maintenance of Clang itself:

> I don't think per-feature LangOptions is viable. There's digit separators, hex float support, binary literal support, different literal suffixes, various keywords, etc and that doesn't seem likely to scale well. Also, it makes it seem like we support a la carte language features when we don't -- we intentionally don't want to let users disable standard features (in general; exceptions exist) and so that approach gives the impression contributors need to check both language mode and feature availability which is a maintenance concern.

How do we avoid this situation, as that's my primary concern.

Reply:
Yeah, I still favor a "DependencyScanning" option over feature specific flags, it seems much simpler to maintain and document in the long run, for exactly the same reason as @AaronBallman